ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/TermCompiler.java

```diff
@@ -0,0 +1,559 @@
+import io.confluent.ksql.execution.interpreter.terms.Term;
+import io.confluent.ksql.execution.util.CoercionUtil;
+import io.confluent.ksql.execution.util.ExpressionTypeManager;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.function.GenericsUtil;
+import io.confluent.ksql.function.KsqlScalarFunction;
+import io.confluent.ksql.function.UdfFactory;
+import io.confluent.ksql.function.types.ArrayType;
+import io.confluent.ksql.function.types.ParamType;
+import io.confluent.ksql.function.types.ParamTypes;
+import io.confluent.ksql.function.udf.Kudf;
+import io.confluent.ksql.schema.ksql.Column;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlArray;
+import io.confluent.ksql.schema.ksql.types.SqlBaseType;
+import io.confluent.ksql.schema.ksql.types.SqlMap;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.DecimalUtil;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.Pair;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@SuppressWarnings("checkstyle:ClassDataAbstractionCoupling")
+public class TermCompiler implements ExpressionVisitor<Term, Void> {
+
+  private final FunctionRegistry functionRegistry;
+  private final LogicalSchema schema;
+  private final KsqlConfig ksqlConfig;
+  private final ExpressionTypeManager expressionTypeManager;
+
+  public TermCompiler(
+      final FunctionRegistry functionRegistry,
+      final LogicalSchema schema,
+      final KsqlConfig ksqlConfig,
+      final ExpressionTypeManager expressionTypeManager
+  ) {
+    this.functionRegistry = functionRegistry;
+    this.schema = schema;
+    this.ksqlConfig = ksqlConfig;
+    this.expressionTypeManager = expressionTypeManager;
+  }
+
+  private Term visitIllegalState(final Expression expression) {
+    throw new IllegalStateException(
+        format("expression type %s should never be visited", expression.getClass()));
+  }
+
+  private Term visitUnsupported(final Expression expression) {
+    throw new UnsupportedOperationException(
+        format(
+            "not yet implemented: %s.visit%s",
+            getClass().getName(),
+            expression.getClass().getSimpleName()
+        )
+    );
+  }
+
+  @Override
+  public Term visitType(final Type node, final Void context) {
+    return visitIllegalState(node);
+  }
+
+  @Override
+  public Term visitWhenClause(final WhenClause whenClause, final Void context) {
+    return visitIllegalState(whenClause);
+  }
+
+  @Override
+  public Term visitInPredicate(
+      final InPredicate inPredicate,
+      final Void context
+  ) {
+    return visitUnsupported(inPredicate);
```

Comment:
for these expressions that aren't supported, are they also not supported in code gen? Seems like `IN` predicate is something that someone may call in the projection

Reply:
I think I had skipped this because I thought that we remove in predicates in favor of `OR` expressions, but you're right that this only happens in the `WHERE` clause.  I added an implementation and a term plus tests.